### ISSUE TITLE
Clarify behavior for denied user consent

### ISSUE THREAD

**Issue by ddorwin** [w3c/encrypted-media#57]
Step 15 of the _Get Supported Configuration_ algorithm defines when to prompt for use of Distinctive Identifier(s). (Issue #51 is related to cleaning this up.) However, it does not say what to do when such consent is denied.

The simple solution is to return null as in step 14.

However, it's possible that the configuration could have been supported in a different way^. We may want to allow reprocessing of the _candidate configuration_ (plus changes based on the denied consent).

We might also want to generalize the text, while still emphasizing the importance for Distinctive Identifiers, to cover checking consent for other reasons (beyond the existing note).

^ For example, if _candidate configuration_'s `distinctiveIdentifier` value is `"optional"`, knowing that the consent is denied might have allowed selection of a different set of `videoCapabilities` that don't require a Distinctive Identifier).

**Comment by mwatson2** [comment]
When did we agree to "recommend" user consent prompts for Distinctive Identifiers ? I thought such prompts were an issue for user agents. Since the identifier is per-site, clearable and encrypted it differs from a cookie only in that it is not easy to steal. For cookies, depending on jurisdiction, it's common simply to provide information to users, rather than seek a priori consent. This should be left open for User Agents to work out the best path, rather than making a recommendation in the specification.

**Comment by mwatson2** [comment]
I opened a new issue (https://github.com/w3c/encrypted-media/issues/63) for the recommendation related to user consent.

This issue can be unblocked to address the question of what should happen if consent is requested and denied. IIUC, one of the points of the site providing multiple options to the requestMediaKeySystemAccess was that it could deal with fallback to lower capabilities for the case where capabilities are unavailable or denies in a single step.

So, the procedures should ensure that when consent is denied, it behaves the same as if that configuration wasn't physically supported. Another requested configuration might still work.

**Comment by ddorwin** [comment]
>IIUC, one of the points of the site providing multiple options to the requestMediaKeySystemAccess was that it could deal with fallback to lower capabilities for the case where capabilities are unavailable or denies in a single step.

The issue here is the combination of user permissions and `optional` along with the various other "optional" parts of the configuration (i.e. `videoCapabilities` entries).

`optional`, the default, means pick whichever value is needed to make this configuration work, and the order means that this configuration is preferred over the next configuration. However, in the algorithm, we wait until the end to check permissions because we only want user agents to request permission when they know the rest of the configuration is supported and the permission is really necessary. Thus, if the permission is denied but the configuration could have been supported, the algorithm does not currently provide the intended behavior.

>So, the procedures should ensure that when consent is denied, it behaves the same as if that configuration wasn't physically supported. Another requested configuration might still work.

That's a good point about behaving the same. It's also an important reason for making the proposed change because it is not currently what is defined by the algorithm. I believe reprocessing the configuration with the new information would provide the desired behavior.

Details: If a Distinctive Identifier was not physically supported, the "If the implementation requires a Distinctive Identifier for any of the combinations in accumulated configuration" block of step 12 would never be entered. In that case, none of the other members (i.e. selected `videoCapabilities` entries) require a Distinctive Identifier. However, when physically supported but requiring permission, an entry requiring 
a Distinctive Identifier could reach step 12. Thus, the configuration that would be selected when a Distinctive Identifier is not physically supported would be skipped if a Distinctive Identifier is physically supported but denied.

**Comment by mwatson2** [comment]
I believe we could address this by splitting _Get Candidate Configuration_ into two. The inner algorithm would contain most of the existing _Get Candidate Configuration_ except that it would handle only the explicit values for `distinctiveIdentifier` (`required` and `not-allowed`).

The outer algorithm would call the inner one directly for the case where `distinctiveIdentifier` is `required` or `not-allowed`. For the case where `distinctiveIdentifier` is `optional` we would first call the inner algorithm with the value `required` and if that fails call it with the value `not-allowed`.

Additionally, I think we should require that the UA persist the user consent status at least within the execution of the `getMediaKeySystemAccess` call, to avoid repeatedly asking the user essentially the same question.

**Comment by mwatson2** [comment]
My proposal above almost works, but not quite: the inner algorithm still needs to take the `optional` value, so that it can upgrade / down based on the result of the audio / video media type processing.

**Comment by mwatson2** [comment]
So, I created a PR #165. I think, technically, this works, but I am not all that happy with it since it repeats the entire selection process with `not-allowed` whenever it fails, independent of the reason for failure. This only really needs to be done if the failure is due to denied consent.

**Comment by mwatson2** [comment]
Ok, it's a little better now. Comments please.

**Comment by ddorwin** [comment]
For reference, this issue was filed against [this version of that section](https://rawgit.com/w3c/encrypted-media/71e4352d4a737f3cdba879217dc9daf74020751d/index.html#get-supported-configuration). Both of the referenced steps have [now](https://rawgit.com/w3c/encrypted-media/27c54136a9a84aaa1bf74d8978d1439ddbac23a3/index.html#get-supported-configuration) been increased to steps 16 and 15, respectively.

**Comment by mwatson2** [comment]
Per the PR review discussion, we may have a more general problem.

The basic 'unit of consent" (in the context of a set of initDataTypes and session types) is a tuple ( distinctive identifier, persistent state, media type, robustness, content type ). Whilst it may seem unlikely that user consent depends on content type, we have delegated all such decisions to the UA, so we have to cope in our algorithms with whatever UAs may decide. For example, codec support may be different with HW and SW DRM as may the privacy properties of distinctive identifiers and thus the need for consent vs simply informing the user.

For each such tuple, before prompting the user, we have three possible statuses:
- consent required
- inform user
- allowed

A consent prompt would ideally collect consent information for multiple of these tuples, either by asking broad questions or, for example, with check boxes to collect more detailed consent information.

After any given consent prompt, the status of any given tuple may be unchanged or may be:
- consent given
- consent denied

I therefore wonder if the _Get consent status_ algorithm should take the full tuple described above and be called from the _Get supported capabilities for audio / video type_ algorithm on a per tuple basis.

However, the actual request for consent should be based on the sum total of tuples that the site is requesting.

So, I wonder if in fact we need a rather different process altogether:
1. construct an ordered list of requested tuples (with all combinations of optional MediaKeyRequirement values and everything else from the MediaKeySystemConfiguration object)
2. prune out those which are not supported, stop if none left
3. evaluate the consent status for every tuple, based on persisted consent data if any
4. prune out any that are "consent denied" based on persisted consent data, stop if none left
5. if the first entry for each media type is not "consent required", return those
6. prompt for consent. The prompt should cover as many tuples as possible.
7. prune out any that are "consent denied", stop if none left
8. got back to step 5

Finally, does `persistedState` really interact with the other tuple elements in the same way as `distinctiveIdentifier` ? First, we could pull out the evaluation of `initDataTypes`, initial processing of `distinctiveIdentifier` and `persistentState` and evaluation of session types to before the above algorithm. The only remaining case where `persistentState` is `optional` is for temporary sessions (otherwise it has been upgraded to `required`).

Presumably, persistent state is not allowed with temporary sessions (by definition, right?). Shouldn't we just downgrade `persistentState` to `not-allowed` at this point and continue ?

**Comment by ddorwin** [comment]
@mwatson2, does my [proposal](https://github.com/w3c/encrypted-media/pull/165/files/#r60819940) address this issue? In it, the user agent applies the restrictions at each step, including selection of codecs, robustness, etc.
<br/>


>Presumably, persistent state is not allowed with temporary sessions (by definition, right?). Shouldn't we just downgrade `persistentState` to `not-allowed` at this point and continue ?

Persistent sessions are one type of persistent state, but we do not limit `persistentState` to only sessions. The implementation expectations for storage, clearing, etc. of all persistent state is the same, though. (See #169.)

**Comment by mwatson2** [comment]
@ddorwin I think your proposal is equivalent to mine, since _restrictions_ is essentially the set of tuples for which consent has been denied. However, your proposal is much easier to implement within the existing text :-)

There may be something in the details, because consent requests would be triggered only at the _Get Supported Configuration and Consent_ and not at the more granular _Get Supported Capabilities for Audio / Video Type_. I'll discover that when I implement it.

**Comment by mwatson2** [comment]
I've updated the PR (#165).

**Comment by paulbrucecotton** [comment]
@ddorwin, @jdsmith3000:  Can you please review Mark's PR (<https://github.com/w3c/encrypted-media/pull/165> to see if it resolves this issue?

**Comment by mwatson2** [comment]
I believe this PR (#165) is ready to merge. If there are no further comments I will do so before our Editors' meeting on Tuesday 5/17.
